### PR TITLE
solution

### PR DESCRIPTION
**Pull request by yelnikov-andrii**
[DEMO LINK](https://yelnikov-andrii.github.io/react_dynamic-list-of-todos/)

### CODE REVIEW

**Review by ZUBOV-ILLIA — APPROVED** [review]
Good Job!

**Comment by Denchepornyuk** [comment]
Screenshot 2022-06-20 at 13 06 54

all buttons with the same user should be selected if one of them selected

File: src/App.tsx [hunk 1]
```diff
@@ -1,25 +1,137 @@
...
 import './styles/general.scss';
 import { TodoList } from './components/TodoList';
 import { CurrentUser } from './components/CurrentUser';
 
 const App: React.FC = () => {
-  const [
-    selectedUserId,
-    // setSelectedUserId,
-  ] = useState(0);
+  const urlTodos = 'https://mate.academy/students-api/todos';
```
Comment (Denchepornyuk) [inline src/App.tsx:11]: move all login with fetch to separate file
create file api/api.ts and move fetch to this file

File: src/App.tsx [hunk 2]
```diff
@@ -1,25 +1,137 @@
...
+
+  const [theTodos, setTheTodos] = useState([{
+    id: 0,
+    createdAt: '',
+    upDatedAt: '',
+    userId: 0,
+    title: '',
+    completed: false,
+  }]);
```
Comment (Denchepornyuk) [inline src/App.tsx:20]:
```suggestion
  const [todos, setTodos] = useState();
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,25 +1,137 @@
...
 import { CurrentUser } from './components/CurrentUser';
 
 const App: React.FC = () => {
-  const [
-    selectedUserId,
-    // setSelectedUserId,
-  ] = useState(0);
+  const urlTodos = 'https://mate.academy/students-api/todos';
+
+  const [theTodos, setTheTodos] = useState([{
```
Comment (Denchepornyuk) [inline src/App.tsx:13]: add type for todos

File: src/App.tsx [hunk 4]
```diff
@@ -1,25 +1,137 @@
...
+    title: '',
+    completed: false,
+  }]);
+
+  useEffect(() => {
+    fetch(urlTodos)
+      .then(response => {
+        if (!response.ok) {
+          throw new Error('Something is wrong');
+        } else {
```
Comment (Denchepornyuk) [inline src/App.tsx:27]: move to api file

File: src/App.tsx [hunk 5]
```diff
@@ -1,25 +1,137 @@
...
+              setTheTodos(todos);
+            });
+        }
+      });
+  }, []);
+
+  const filterTodos = (inputedValue: string) => {
+    fetch(urlTodos)
+      .then(response => {
+        response.json()
```
Comment (Denchepornyuk) [inline src/App.tsx:39]: move to api file

File: src/App.tsx [hunk 6]
```diff
@@ -1,25 +1,137 @@
...
+          .then(todos => {
+            const filtered = [...todos].filter((todo) => {
+              return todo.title.includes(inputedValue);
```
Comment (Denchepornyuk) [inline src/App.tsx:42]: don't forget to handle lower and upper cases

File: src/App.tsx [hunk 7]
```diff
@@ -1,25 +1,137 @@
...
+                return todo;
+              }
+
+              if (selectedInput === 'active') {
+                return todo.completed === false;
+              }
+
+              if (selectedInput === 'completed') {
+                return todo.completed === true;
+              }
```
Comment (Denchepornyuk) [inline src/App.tsx:66]: use switch case

File: src/App.tsx [hunk 8]
```diff
@@ -1,25 +1,137 @@
...
+          });
+      });
+  };
+
+  const selectTodos = (selectedInput: string) => {
+    fetch(urlTodos)
+      .then(response => {
+        response.json()
+          .then(todos => {
+            const selected = [...todos].filter((todo) => {
```
Comment (Denchepornyuk) [inline src/App.tsx:55]: you don't have to make an array copy, filter is immutable method

File: src/App.tsx [hunk 9]
```diff
@@ -1,25 +1,137 @@
...
+    }
+
+    return array;
+  }
+
+  const randomSort = () => {
+    fetch(urlTodos)
+      .then(response => {
+        response.json()
+          .then(todos => {
```
Comment (Denchepornyuk) [inline src/App.tsx:96]: move to api file

File: src/components/CurrentUser/CurrentUser.tsx [hunk 10]
```diff
@@ -1,12 +1,70 @@
...
+};
 
-    <h3 className="CurrentUser__name">Ervin Howell</h3>
-    <p className="CurrentUser__email">Shanna@melissa.tv</p>
-    <p className="CurrentUser__phone">010-692-6593 x09125</p>
-  </div>
-);
+export const CurrentUser: React.FC <Props> = ({ selectedUserId, onClear }) => {
+  const userId = selectedUserId;
+  const urlUser = `https://mate.academy/students-api/users/${userId}`;
```
Comment (Denchepornyuk) [inline src/components/CurrentUser/CurrentUser.tsx:12]: move to api file

File: src/components/CurrentUser/CurrentUser.tsx [hunk 11]
```diff
@@ -1,12 +1,70 @@
...
+    },
+  );
+
+  useEffect(() => {
+    fetch(urlUser)
+      .then(response => {
+        if (!response.ok) {
+          setUser({
+            id: 0, name: '', email: '', phone: '',
+          });
```
Comment (Denchepornyuk) [inline src/components/CurrentUser/CurrentUser.tsx:26]: move to api file

File: src/components/CurrentUser/CurrentUser.tsx [hunk 12]
```diff
@@ -1,12 +1,70 @@
...
-);
+export const CurrentUser: React.FC <Props> = ({ selectedUserId, onClear }) => {
+  const userId = selectedUserId;
+  const urlUser = `https://mate.academy/students-api/users/${userId}`;
+
+  const [user, setUser] = useState(
+    {
+      id: 0, name: '', email: '', phone: '',
+    },
+  );
```
Comment (Denchepornyuk) [inline src/components/CurrentUser/CurrentUser.tsx:18]:
```suggestion
  const [user, setUser] = useState(null);
```
user can be null, add type for user

File: src/components/CurrentUser/CurrentUser.tsx [hunk 13]
```diff
@@ -1,12 +1,70 @@
...
+      });
+  }, [userId]);
+
+  return (
+    <div className="CurrentUser">
+      {user.id === 0 && user.name === ''
+      && user.email === '' && user.phone === '' ? ('User not found') : (
+          // eslint-disable-next-line react/jsx-indent
+          <>
+            <h2 className="CurrentUser__title"><span>{`Selected user: ${user.id}`}</span></h2>
```
Comment (Denchepornyuk) [inline src/components/CurrentUser/CurrentUser.tsx:47]: don't write code in one line

File: src/components/CurrentUser/CurrentUser.tsx [hunk 14]
```diff
@@ -1,12 +1,70 @@
...
+              }
+            });
+        }
+      });
+  }, [userId]);
+
+  return (
+    <div className="CurrentUser">
+      {user.id === 0 && user.name === ''
+      && user.email === '' && user.phone === '' ? ('User not found') : (
```
Comment (Denchepornyuk) [inline src/components/CurrentUser/CurrentUser.tsx:44]: check only user instead of all these fields

File: src/components/TodoList/TodoList.tsx [hunk 15]
```diff
@@ -1,44 +1,146 @@
...
+  const [inputedValue, setInputedValue] = useState('');
+
+  return (
+    <div className="TodoList">
+      <form className="form">
+        <input
+          placeholder="input the title"
+          data-cy="filterByTitle"
+          className="form__input"
+          value={inputedValue}
```
Comment (Denchepornyuk) [inline src/components/TodoList/TodoList.tsx:45]: this form shoudn't be in TodoList component

File: src/components/TodoList/TodoList.tsx [hunk 16]
```diff
@@ -1,44 +1,146 @@
...
+            value="completed"
+          >
+            Completed
+          </option>
+        </select>
+
+        <button
+          type="button"
+          className="btn"
+          onClick={onSorted}
```
Comment (Denchepornyuk) [inline src/components/TodoList/TodoList.tsx:80]: use onSubmit in form instead of onClick in button

**Review by Denchepornyuk — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 17]
```diff
@@ -1,25 +1,166 @@
...
+        });
+
+        setTheTodos(filtered);
+      });
+  };
+
+  const selectTodos = (selectedInput: string) => {
+    request(urlTodos)
+      .then(todos => {
+        const selected = [...todos].filter((todo) => {
```
Comment (Denchepornyuk) [inline src/App.tsx:44]: you don't have to make an array copy, filter is immutable method

**Review by Denchepornyuk — APPROVED** [review]

**Pull request closed without merging**